tests/cpp/test_resharding.cpp

```diff
@@ -291,6 +279,196 @@ TEST_F(ReshardingTest, Add_InputsParallelizedDifferently) {
   EXPECT_TRUE(isResharding(z->definition()));
 }
 
+TEST_F(ReshardingTest, Add_Broadcast) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  TensorView* x = makeContigTensor(2);
+  x->setDeviceMesh({0, 1});
+  TensorView* y = makeContigTensor(1);
+  y->setDeviceMesh({0, 1});
+  y = broadcast(y, {true, false});
+  TensorView* z = add(x, y);
+
+  for (auto* tv : {x, y, z}) {
+    tv->axis(0)->parallelize(ParallelType::DIDx);
+  }
+
+  EXPECT_FALSE(isResharding(z->definition()));
+}
+
+TEST_F(ReshardingTest, Matmul_NoResharding) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  TensorView* x = makeContigTensor(2);
+  TensorView* y = makeContigTensor(2);
+  TensorView* z = matmul(x, y);
+
+  for (auto* tv : {x, y, z}) {
+    tv->setDeviceMesh({0, 1});
+  }
+  for (auto* tv : {x, z}) {
+    tv->axis(0)->parallelize(ParallelType::DIDx);
+  }
+
+  EXPECT_FALSE(isResharding(z->definition()));
+}
+
+TEST_F(ReshardingTest, Matmul_Resharding) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  TensorView* x = makeContigTensor(2); // [iM, iK]
+  TensorView* y = makeContigTensor(2); // [iK, iN]
+  TensorView* z = matmul(x, y); // [iM, iN, rk]
+
+  for (auto* tv : {x, y, z}) {
+    tv->setDeviceMesh({0, 1});
+  }
+  x->axis(0)->parallelize(ParallelType::DIDx);
+  y->axis(1)->parallelize(ParallelType::DIDx);
+  z->axis(0)->parallelize(ParallelType::DIDx);
+
+  // iN is sharded in y but not in z.
+  EXPECT_TRUE(isResharding(z->definition()));
+}
+
+TEST_F(ReshardingTest, Allgather) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  constexpr int64_t kNumDevices = 2;
+  TensorView* in = makeContigTensor(2);
+  in->setDeviceMesh(DeviceMesh::createForNumDevices(kNumDevices));
+  TensorView* out = set(in);
+
+  in->split(0, kNumDevices, /*inner_split=*/false);
+  in->axis(0)->parallelize(ParallelType::DIDx);
+
+  EXPECT_TRUE(isResharding(out->definition()));
+}
+
+TEST_F(ReshardingTest, ReduceScatter) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  constexpr int64_t kNumDevices = 2;
+  const auto mesh = DeviceMesh::createForNumDevices(kNumDevices);
+
+  TensorView* in = makeContigConcreteTensor({6, 10});
+  in->split(0, kNumDevices, /*inner_split=*/false);
+
+  TensorView* rfactor = reshape(
+      in,
+      {in->axis(0)->extent(), in->axis(1)->extent(), in->axis(2)->extent()});
```

Comment:
@naoyam I believe there's a better way to write it. commitLeafToLogical?

Reply:
I'm not entirely sure what is supposed to be computed here. Looks like only the outermost dimension of `rfactor` is reduced. What is assumed about the input? What is the output?

Reply:
I updated the test so it's no longer the concern. 

FWIW, a more realistic example is the reduce-scatter after a sequence-parallel MLP forward. With DID loop split, it'll look like

```
in: logical=[B, S, 4E], loop=[D, B, S, 4E/D]
w: logical=[E, 4E], loop=[D, E, 4E/D]
out=linear(in,w): logical=[B, S, E], loop=[D, B, S/D, E]
```

However, `out->definition()` does both linear and resharding, so it has to be decomposed before lowering. 

I *think* we'll end up decomposing it as:
```
in: logical=[B, S, 4E], loop=[D, B, S, 4E/D]
in': logical/loop=[D, B, S, 4E/D]
w: logical=[E, 4E], loop=[D, E, 4E/D]
w': logical/loop=[D, E, 4E/D]
local_out=linear(in',w'): logical/loop=[D, B, S, E]
out=sum(local_out): logical=[B, S, E], loop=[D, B, S/D, E]
```

Reply:
I wonder why reshape is used, but this is probably a topic for a future PR.

Reply:
> I wonder why reshape is used, but this is probably a topic for a future PR.

Because I can't find a better way at this moment... `sum` (which represents ReduceScatter) has to reduce a device dimension in the **logical** domain. Therefore, the linear has to produce a device dimension in its **logical** domain, and the reshape.

Reply:
This sounds like `TensorView::rFactor`.